File: CPP/ByRound/1845/1845_F_Swimmers_in_the_Pool.cpp
// Problem: CF 1845 F - Swimmers in the Pool
// https://codeforces.com/contest/1845/problem/F

/*
 * Problem: F. Swimmers in the Pool
 * 
 * Purpose: Calculate the number of meeting moments when swimmers are at the same position in a pool,
 *          considering they swim back and forth. This usesFFT-based convolution and inclusion-exclusion
 *          principles to count overlapping swimmer positions efficiently.
 * 
 * Algorithms/Techniques:
 *  - Fast Fourier Transform (FFT) with NTT (Number Theoretic Transform) for polynomial multiplication
 *  - Inclusion-exclusion principle to avoid overcounting
 *  - Modular arithmetic for large numbers
 * 
 * Time Complexity: O(N log N + S log S), where N is the number of distinct speeds (up to 2e5),
 *                   and S is the size of the problem domain (up to 2e5).
 * Space Complexity: O(N + S), for storing arrays of size up to 2^19 and S.
 */

#include <iostream>
#include <utility>

using namespace std;
typedef long long ll;
const int N = 1 << 19, S = 2e5 + 4;
int l, t, n, a[N], A[N], B[N], rv[N];
const int M = 998244353, G = 3, I = 332748118;
int myp(int x, int t) {
  int a = 1;
  for (; t; t >>= 1, x = (ll)x * x % M)
    if (t & 1)
      a = (ll)a * x % M;
  return a;
}
void ntt(int *A, int z) {
  for (int i = 0; i < N; i++)
    if (i < rv[i])
      swap(A[i], A[rv[i]]);
  for (int t = 0; t < 19; t++) {
    int W = myp(~z ? G : I, (M - 1) >> (t + 1)), T = 1 << t;
    for (int i = 0; i < N; i += T + T) {
      for (int j = 0, Wn = 1; j < T; j++, Wn = (ll)Wn * W % M) {
        int x = A[i + j], y = (ll)A[i + T + j] * Wn % M;
        A[i + j] = x + y < M ? x + y : x + y - M,
              A[i + T + j] = x - y < 0 ? x - y + M : x - y;
      }
    }
  }
  if (z < 0) {
    int iv = myp(N, M - 2);
    for (int i = 0; i < N; i++)
      A[i] = (ll)A[i] * iv % M;
  }
}
const int P = 1e9 + 7;
bool ok[S * 2];
int dp[S * 2], ans;
int main() {
  cin >> l >> t >> n, l *= 2; // Double length because swimming back and forth
  for (int i = 0; i < n; i++)
    cin >> a[i]; // Read swimmer speeds
  for (int i = 0; i < n; i++)
    A[a[i]]++, B[200000 - a[i]]++; // Use two arrays: one for speed counts, one for reverse speeds
  for (int i = 1; i < N; i++)
    rv[i] = (rv[i >> 1] >> 1) + ((i & 1) << 18); // Precompute bit-reversal indices
  ntt(A, 1), ntt(B, 1); // Apply NTT to both arrays
  for (int i = 0; i < N; i++)
    B[i] = (ll)B[i] * A[i] % M, A[i] = (ll)A[i] * A[i] % M; // Multiply A*A and A*B
  ntt(A, -1), ntt(B, -1); // Perform inverse NTT to get the result
  for (int i = 0; i < n; i++)
    A[a[i] * 2]--; // Subtract self-interactions
  for (int i = 1; i < S * 2; i++)
    if (A[i])
      ok[i] = 1; // Mark valid intervals where overlaps occur
  for (int i = 1; i < S; i++)
    if (B[200000 + i])
      ok[i] = 1; // Check for valid positions based on reverse speeds
  for (int i = 1; i < S * 2; i++)
    for (int j = i + i; j < S * 2; j += i)
      ok[i] |= ok[j]; // Propagate valid flags using inclusion-exclusion
  for (int i = 1; i < S * 2; i++)
    if (ok[i]) {
      if ((dp[i] += (ll)t * i / l % P) >= P)
        dp[i] -= P;
      for (int j = i + i; j < S * 2; j += i)
        if ((dp[j] -= dp[i]) < 0)
          dp[j] += P;
      if ((ans += dp[i]) >= M)
        ans -= P;
    }
  cout << ans; // Output the final answer modulo M
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1845/1845_F_Swimmers_in_the_Pool.py
# Problem: CF 1845 F - Swimmers in the Pool
# https://codeforces.com/contest/1845/problem/F

"""
Algorithm: Number Theoretic Transform (NTT) + Inclusion-Exclusion Principle

Purpose:
This code calculates the number of meeting moments among swimmers in a pool.
Each swimmer swims back and forth between points 0 and l with constant speed.
We need to find how many times at least two swimmers are at the same point in the pool during the time t.

Approach:
1. Use NTT to efficiently compute convolution of swimmer positions.
2. Identify points where at least two swimmers meet using convolution results.
3. Apply inclusion-exclusion principle to count actual meeting moments within [0, t].

Time Complexity: O(n log n + M log M), where M = max speed * 2
Space Complexity: O(M), for arrays used in NTT and for storing intermediate results.

Where:
- n is the number of swimmers
- l is the pool length
- t is the duration of the process
- M = max speed * 2

"""

import sys

input = lambda: sys.stdin.readline().rstrip()


def II(base=10):
    return int(input(), base)


def LI():
    return list(map(int, input()))


def LII():
    return list(map(int, input().split()))


mod = 998244353
G = 3
r = [pow(3, mod >> i, mod) for i in range(23)]
ir = [pow(i, mod - 2, mod) for i in r]


def ntt(A, N, sgn=1):
    L = N.bit_length() - 1
    rev = [0] * N
    for i in range(N):
        j = rev[i] = (rev[i >> 1] >> 1) | (i & 1) * (N >> 1)
        if i < j:
            A[i], A[j] = A[j], A[i]
    for i in range(L):
        a = 1 << i
        step = r[i + 1] if sgn == 1 else ir[i + 1]
        for j in range(0, N, a * 2):
            w = 1
            for k in range(j, j + a):
                x, y = A[k], A[k + a] * w % mod
                A[k], A[k + a] = (x + y) % mod, (x - y) % mod
                w = w * step % mod
    if sgn == -1:
        inv = pow(N, -1, mod)
        for i, x in enumerate(A):
            A[i] = x * inv % mod


l, t = LII()
n = II()
A = LII()
m = max(A)
N = 1 << (m * 2 + 1).bit_length()
f = [0] * N
B, B2 = [0] * N, [0] * N
for a in A:
    B[a] = 1
    B2[m - a] = 1
ntt(B, N)
ntt(B2, N)
C = [b * b for b in B]
C2 = [b * b2 for b, b2 in zip(B, B2)]
ntt(C, N, -1)
ntt(C2, N, -1)
for a in A:
    C[a * 2] -= 1
for i in range(1, N):
    if C[i] or (i + m < N and C2[i + m]):
        f[i] = 1
h = [[] for _ in range(N)]
for i in range(1, N):
    for j in range(i * 2, N, i):
        h[j].append(i)
for i in range(N - 1, 0, -1):
    if f[i]:
        for j in h[i]:
            f[j] = 1

mod = 10**9 + 7
res = 0
g = [0] * N
for i in range(1, N):
    if f[i]:
        g[i] = (t * i) // (2 * l)
        for j in h[i]:
            g[i] -= g[j]
            g[i] %= mod
        res += g[i]
        res %= mod
print(res)


# https://github.com/VaHiX/CodeForces/